blog/2019-10-08-handling-unchanged-postgres-toast-values.adoc

```diff
@@ -0,0 +1,316 @@
+When processing the second event, the stream processing application will not be able to look up the `biography` value stored earlier on, as it has been under the old key.
+
+One way to address this would be to expose the original key value e.g. as a message header of the insert event.
+This requirement is tracked as https://issues.jboss.org/browse/DBZ-1531[DBZ-1531];
+let us know if you'd like to contribute and implement this feature.
+===============================
+
+== When to Use What?
+
+We've discussed different options for dealing with unchanged TOAST column values in Debezium's data change events.
+Which one should be used in which case then?
+
+Changing the replica identity to `FULL` is the easiest approach by far:
+a single configuration to the source table avoids the problem to begin with.
+It's not the most efficient solution, though, and some DBAs might be reluctant to apply this setting.
+
+When using the change events to update some kind of sink data store,
+it may sound attractive at first to simply omit any field with the special marker value when issuing an update.
+But this technique has some downsides: not all data stores and the corresponding connectors might support partial updates.
+Instead there might only be the option to do full updates to a record in the sink data store based on the incoming data.
+Even when that option exists, it might be sub-optimal.
+E.g. for a SQL database, a statement just with the available values may be executed.
+This is at odds with efficient usage of prepared statements and batching, though:
+as the "shape" of the data may change between two updates to the same table,
+the same prepared statement cannot be re-used and performance may suffer.
+
+The trigger-based approach isn't prone to these problems:
+any updates to a table will have the same number of columns, so the consumer (e.g. a sink connector) may re-use the same prepared statement and batch multiple records into a single execution.
+One thing to be aware of is the organizational cost associated with this approach:
+triggers must be installed for each affected column and be kept in sync when table structures change.
+This must be done individually in each sink datastore, and not all stores have may have support for triggers to begin with.
+But where possible, triggers can be a great solution.
+
+Finally, stream processing makes the usage of TOAST-able columns and the absence of their values in update events fully transparent to consumers.
+The enrichment logic is implemented in a single place, from which all the consumers of the change event stream benefit,
+without the need for individual solutions in each one of them.
+Also, it's the only viable solution if consumers themselves are stateless and don't have any way to materialize the last value of such column, e.g. when streaming change events to a browser via web sockets or GraphQL subscriptions.
+The price to pay is the overhead of maintaining and operating a separate service.
```

Comment:
I think it would be good to emphasize that the demo code can be generalized for arbitrary event and it is not necessary to have a specific code for each table/colunm.

Reply:
Yes, I've added something for that 👍